components/templates/src/global_fns/load_data.rs

```diff
@@ -0,0 +1,346 @@
+    }
+    else if let Some(url) = url_arg {
+        return Url::parse(&url).map(|parsed_url| ProvidedArgument::URL(parsed_url)).map_err(|e| format!("Failed to parse {} as url: {}", url, e).into());
+    }
+
+    return Err(GET_DATA_ARGUMENT_ERROR_MESSAGE.into());
+}
+
+fn read_data_file(base_path: &PathBuf, full_path: PathBuf) -> Result<String> {
+    if !is_path_in_directory(&base_path, &full_path).map_err(|e| format!("Failed to read data file {}: {}", full_path.display(), e))? {
+        return Err(format!("{} is not inside the base site directory {}", full_path.display(), base_path.display()).into());
+    }
+    return read_file(&full_path)
+        .map_err(|e| format!("`load_data`: error {} loading file {}", full_path.to_str().unwrap(), e).into());
+}
+
+fn get_output_format_from_args(args: &HashMap<String, Value>, provided_argument: &ProvidedArgument) -> Result<String> {
+    let format_arg = optional_arg!(
+        String,
+        args.get("format"),
+        "`load_data`: `format` needs to be an argument with a string value, being one of the supported `load_data` file types (csv, json, toml)"
+    );
+
+    if let Some(format) = format_arg {
+        return Ok(format);
+    }
+    return match provided_argument {
+        ProvidedArgument::PATH(path) => path.extension().map(|extension| extension.to_str().unwrap().to_string()).ok_or(format!("Could not determine format for {} from extension", path.display()).into()),
+        _ => Ok(String::from("plain"))
+    }
+}
+
+
+/// A global function to load data from a data file.
+/// Currently the supported formats are json, toml and csv
+pub fn make_load_data(content_path: PathBuf, base_path: PathBuf) -> GlobalFn {
+    let mut headers = header::HeaderMap::new();
+    headers.insert(header::USER_AGENT, format!("{} {}", crate_name!(), crate_version!()).parse().unwrap());
+    let client = Arc::new(Mutex::new(Client::builder().build().expect("reqwest client build")));
+    let result_cache: Arc<Mutex<HashMap<String, Value>>> = Arc::new(Mutex::new(HashMap::new()));
+    Box::new(move |args| -> Result<Value> {
+        let provided_argument = get_data_from_args(&content_path, &args)?;
+
+        let file_format = get_output_format_from_args(&args, &provided_argument)?;
+
+        let cache_key = get_cache_key(&provided_argument, &file_format);
+
+        let mut cache = result_cache.lock().expect("result cache lock");
+        let response_client = client.lock().expect("response client lock");
+        if let Some(cached_result) = cache.get(&cache_key) {
+            return Ok(cached_result.clone());
+        }
+
+        let data = match provided_argument {
+            ProvidedArgument::PATH(path) => read_data_file(&base_path, path),
+            ProvidedArgument::URL(url) => {
+                let mut response = response_client.get(url.as_str()).send().and_then(|res| res.error_for_status()).map_err(|e| format!("Failed to request {}: {}", url, e.status().expect("response status")))?;
```

Comment:
URL requests should add accept headers to the request since some sites send content following it.